platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,290 @@
+nodes.
+The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address
+book.
+
+This proposal specifies that Roster(s) will be created from the CAB by the Hedera app and passed to the platform code.
+The mechanism for doing so is detailed below.
+
+### Requirements
+
+- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
+  copying.
+- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
+  platform to manage the lifecycle of submitted rosters.
+- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
+  Event Validation etc.) must be adapted to use rosters instead.
+- Address book paradigm discontinued within the platform codebase, only use rosters instead
+
+### Architecture and/or Components
+
+###### Roster API
+
+The Hedera app is already responsible for managing the address book.
+We propose that it continues to do so, and at adoption time, creates a candidate Roster object from the CAB and set it
+in the state.
+The State will contain one map of rosters, keyed by roster’s hash - ```Map<Hash<Roster>, Roster>```.
+It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
+This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, the optional
+previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.
+There will be new fields in PlatformState - `candidateRosterHash` and `activeRosterHash` - such that at adoption time,
+the way to trigger the adoption of a roster will be by the client code inserting a roster in the roster map, alongside
+setting the `candidateRosterHash` field in the PlatformState.
+The `activeRosterHash` will be private to the platform and will not be settable or modifiable from outside the platform.
+
+This indirection avoids moving data around in the merkle state which requires data to be copied into the block stream,
+which is computationally expensive.
+Another benefit of this approach is that adoption trigger becomes very simple (app sets the roster) with delineated
+responsibilities between the app and the platform.
+
+Some edge cases worth considering include:
+
+1. multiple concurrent roster submission - has a potential for introducing race conditions. We will prevent this by
+   guaranteeing immutability on rosters.
+2. Size control on the map of rosters - we certainly don’t want this map to grow infinitely so insertion of rosters will
+   be controlled, with clear rules for removal of unused rosters (e.g. the acceptance of a new Candidate Roster
```

Comment:
How is this enforced?

Reply:
I think the way the map can keep growing is by rapidly adding candidate rosters so fast (likely) or adopting new rosters rapidly (unlikely). As the example says, the quickest way to prevent  this is to attack the likely scenario - remove the current candidate roster once a new one is hashed. It will also address @kelly's question about why we need previous candidate roster. I've updated this to make this point very clear.